python/prima.py

```diff
@@ -0,0 +1,341 @@
+import ctypes as ct
+import ctypes.util
+import os
+from dataclasses import dataclass
+
+
+_lib_name = "primac"
+try:
+    # import from current dir
+    import numpy as np
+
+    _path = os.path.abspath(os.path.dirname(__file__))
+    _handle = np.ctypeslib.load_library(_lib_name, _path)
+except Exception:
+    # else rely on loader env variables
+    _lib_loc = ctypes.util.find_library(_lib_name)
+    if _lib_loc is None:
+        raise ValueError(f"Cannot find {_lib_name} library")
+    _handle = ct.cdll.LoadLibrary(_lib_loc)
+
+# primac callback form
+_prima_function_callback = ct.CFUNCTYPE(
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double)
+)
+_prima_function_con_callback = ct.CFUNCTYPE(
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double),
+    ct.POINTER(ct.c_double)
+)
+
+# prima_get_rc_string returns char*
+_handle.prima_get_rc_string.restype = ct.c_char_p
+
+# pure Python function converters
+
+
+def _prima_pure_callback(func, n):
+    def _inner(cx, cf):
+        x = [cx[i] for i in range(n)]
+        f = func(x)
+        cf[0] = f
+
+    return _prima_function_callback(_inner)
+
+
+def _prima_pure_con_callback(func, n, m):
+    def _inner(cx, cf, ccon):
+        x = [cx[i] for i in range(n)]
+        f, con = func(x)
+        cf[0] = f
+        for i in range(m):
+            ccon[i] = con[i]
+
+    return _prima_function_con_callback(_inner)
+
+
+@dataclass
+class Result:
+    """Optimization result."""
+    x: list
```

Comment:
Is there any reason to use a `list` instead of the standard `numpy.ndarray`? This would then provide a similar structure to [OptimizeResult](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html).

Reply:
we dont need to depend on numpy, beside a list can be transformed into ndarray